File: common/src/main/java/juuxel/adorn/client/FluidRenderingBridge.java
package juuxel.adorn.client;

import juuxel.adorn.fluid.FluidReference;
import juuxel.adorn.util.InlineServices;
import juuxel.adorn.util.Services;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.network.chat.Component;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.BlockAndTintGetter;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@InlineServices
public interface FluidRenderingBridge {
    @Nullable TextureAtlasSprite getStillSprite(FluidReference volume);

    int getColor(FluidReference volume, @Nullable BlockAndTintGetter world, @Nullable BlockPos pos);

    default int getColor(FluidReference volume) {
        return getColor(volume, null, null);
    }

    boolean fillsFromTop(FluidReference volume);

    List<Component> getTooltip(FluidReference volume, TooltipFlag type, @Nullable Integer maxAmountInLitres);

    @InlineServices.Getter
    static FluidRenderingBridge get() {
        return Services.load(FluidRenderingBridge.class);
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/FluidRenderingBridge.kt
package juuxel.adorn.client

import juuxel.adorn.fluid.FluidReference
import net.minecraft.client.item.TooltipContext
import net.minecraft.client.texture.Sprite
import net.minecraft.text.Text
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockRenderView

interface FluidRenderingBridge {
    fun getStillSprite(volume: FluidReference): Sprite?

    fun getColor(volume: FluidReference, world: BlockRenderView? = null, pos: BlockPos? = null): Int

    fun fillsFromTop(volume: FluidReference): Boolean

    fun getTooltip(volume: FluidReference, context: TooltipContext, maxAmountInLitres: Int?): List<Text>

    companion object {
        fun get() = ClientPlatformBridges.get().fluidRendering
    }
}
